### PR TITLE
Melee damage while wearing filthy clothing may result in infection

### PR DESCRIPTION
**Pull request by Night-Pryanik**
Also, water needed for washing now depend on item volume. 

Originated from #20160, which, in turn, originated from #17591. Based on @mugling 's work and @Coolthulhu 's suggestions.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Again: I'd start with this being optional.
Whether opt-in or opt-out, it shouldn't be "no-opt".

The problem with it is that it is a lot of added hassle (player cost) for relatively little character cost. Simulationists may like it, but a lot of people hated the washing aspect of original filthy clothing and the extra nerf to already bad zombie-scavenged armor. It made scavenging unappealing, instead making it even more about the mass crafting.

Ideas that attract as much hostility as having to wash clothing in the game should be either optional or not implemented.

As for the implementation:
The function you have there is a bit too binary. Wearing 3 layers of thick filthy clothing counts the same as wearing one filthy fanny pack.
Would be better if you scaled the infection chance by total coverage of the items on the slot. That is, sum up the coverage of items covering the part (could go over 100%), then multiply it by damage.
The repeated applications don't stack well. Zombie bites after the first increase the bite timer, same could go here.
Finally, bite and infection effects do not work well together. It's one or the other. Either make bite counter go to infection if both are present in `player::hardcoded_effects`, special case it here or make it not happen when the player already has an infection.

**Comment by Night-Pryanik** [comment]
> The repeated applications don't stack well. Zombie bites after the first increase the bite timer, same could go here.

Why? Is it a performance issue?

**Comment by Night-Pryanik** [comment]
> Finally, bite and infection effects do not work well together. It's one or the other. Either make bite counter go to infection if both are present in player::hardcoded_effects, special case it here or make it not happen when the player already has an infection.

But infection from filthy clothing isn't `effect_infected`, it's just `effect_bite`. So being bitten by zombie and being infected from filthy clothing is just the same. Do two bite effects from different sources one after another work well together?

**Comment by Coolthulhu** [comment]
> Why? Is it a performance issue?

No, the problem is that you apply the bite effect with time 1. This means that if you get infected once, further infections from dirt only increase the timer by 1 turn, unlike zombie bites which increase it by 250 turns or so.

> But infection from filthy clothing isn't effect_infected, it's just effect_bite.

That's part of the problem: zombie bites first cause bite, then if the bite progresses to infected, they worsen the infection rather than reapplying the (now removed) bite effect.
The game won't work well if the same character has one part at the same bitten and infected - when a part becomes infected, it should not get the bite effect until it stops being infected.
Look at how zombies apply bite.

**Comment by Night-Pryanik** [comment]
> No, the problem is that you apply the bite effect with time 1. This means that if you get infected once, further infections from dirt only increase the timer by 1 turn, unlike zombie bites which increase it by 250 turns or so.

I meant why the timer is needed in first place.

> The game won't work well if the same character has one part at the same bitten and infected - when a part becomes infected, it should not get the bite effect until it stops being infected.

So it's pre-existing problem, and is not added by infection-chance-from-melee-hit mechanic?

**Comment by keyspace** [comment]
Shouldn't this be part of `Character`, not `player`, to include NPCs?..

**Comment by Coolthulhu** [comment]
> I meant why the timer is needed in first place.

So that bite can advance to infection.

> So it's pre-existing problem, and is not added by infection-chance-from-melee-hit mechanic?

Technically yes, but it doesn't happen in unmodded game. You can fix it or work around it, but you can't just apply bite without taking care of infected.

**Comment by kevingranade** [comment]
On Jan 30, 2017 11:34 AM, "Coolthulhu" <notifications@github.com> wrote:

Again: I'd start with this being optional.
Whether opt-in or opt-out, it shouldn't be "no-opt".

...

Ideas that attract as much hostility as having to wash clothing in the game
should be either optional or not implemented.

"Hostility-based design" is about as appealing as, "popularity based
design", i.e. not at all.

If I cared about popularity of features in any way I never would have
started the dda fork, I don't know why you think that would change now.

Increased infection chance on hit was a feature you were requesting in
#17254, what changed?

**Comment by Barhandar** [comment]
>Increased infection chance on hit was a feature you were requesting in #17254, what changed?

Increased infection chance on hit is not related to filthy clothes, and you know this.
> "Hostility-based design" is about as appealing as, "popularity based design", i.e. not at all.

Which is not a reason to reject the base complaints. And the base complaint is thusly:
>The problem with it is that it is a lot of added hassle (player cost) for relatively little character cost. Simulationists may like it, but a lot of people hated ~~the washing aspect of original filthy clothing and~~ the extra nerf to already bad zombie-scavenged armor. It made scavenging unappealing, instead making it even more about the mass crafting.

I agree with that complaint. It's an added hassle and yet another carpet-nerf to things that shouldn't be carpet-nerfed. And I don't think CDDA is about "realism" or "simulation" or other such bollocks, so "this is realistic!" argument falls entirely flat.

**Comment by Night-Pryanik** [comment]
> And I don't think CDDA is about "realism" or "simulation" or other such bollocks

Simulating oh so many aspects of real life, so what CDDA is about then?

**Comment by CoroNaut** [comment]
To be honest, we have a community that varies quite widely on how they envision their cataclysm. Also, we have lost some of the rate at which we are making issues or suggestions after the engine debacle, and even more so after it was removed. This is sad that the community was this torn over something simple similar to this.

Ideas only attract "hostility" because people disagree with the idea which means that it is unpopular. This is basically the opposite to the "popularity based design". Therefore I see no reason to reason to call it "Hostility-based design". The fact that some players dislike already in-place additions is exactly the same as disliking it before it gets added except that the player has the ability to disapprove.

Also, we have a community going here and it is slowly getting smaller solely because of hostility in recent events. Having any voting system typically means that it was made because people take both sides, perhaps we can come to some conclusion that everyone likes on these issues. 

This issue in particular seems like it could work. Having filthy clothing+having a cut=possiblity of getting infected. There shouldn't be any hostility over something like this to be had. I may not like that the game gets a smidgen more difficult, but it still adds something that is plausible into the game. If we just deny every suggestion that surfaces, we will just start to stagnate.

Whether this game is realistic, a simulation or scifi should really be cut out clear as day as well.

**Comment by Coolthulhu** [comment]
> Increased infection chance on hit was a feature you were requesting in #17254, what changed?

My idea about infection from clothing was mostly a way to salvage the whole dirty clothing thing. I was thinking "how to improve filthy clothing" not "how to make filthy clothing an actually good mechanic".
Then I saw the arguments against it:
* It doesn't actually solve the problem of unneeded nerfs to zombie-dropped clothing
* It stresses the whole "craft to win" start game tactic, where unskilled non-tailors start the first game day by ignoring hunger and thirst and just crafting clothing
* It may be worse than the original morale penalty filthiness precisely because of what was supposed to make it good: because it matters and can't be avoided by dropping the clothing during downtime

You spoke against salvaging ideas/PRs multiple times. I admit that in the specific case of filthy clothing, my attempts to salvage it were wrong. I had a "better" idea, but better isn't always good.

Now that realism was brought to discussion, I have to remind everyone:
It isn't realism when you simulate something so much that other aspects completely fall apart. Filthy clothing getting attention is anti-realistic in the end, because it only results in everyone having to be a skilled tailor, as useful non-filthy clothing is rarer than hard drugs and guns ('MERICA!).
It is only as realistic as the actual effects not just on gameplay, but on game style. "Cataclysm" or "Craftaclysm"?

**Comment by kevingranade** [comment]
* It doesn't actually solve the problem of unneeded nerfs to zombie-dropped clothing

Too many double negatives, do you mean zombie drops don't need nerfs? (they could definitely use some, they're everywhere, see below for fuller argument) Or do you mean that this isn't actually a nerf?
* It stresses the whole "craft to win" start game tactic, where unskilled non-tailors start the first game day by ignoring hunger and thirst and just crafting clothing

That's a problem with crafting then, scavenging from houses and shops should (according to the general scenario) the dominant way to acquire equipment.  Zombie drops shouldn't be encouraged either, repairing and cleaning zombie drops seems like an extreme last-resort type of thing to do, if it's more attractive than looting houses or stores, we have seriously messed up incentives.
* It may be worse than the original morale penalty filthiness precisely because of what was supposed to make it good: because it matters and can't be avoided by dropping the clothing during downtime

So basically it does what it's intended to do, but for metagame reasons completely unrelated to the feature, you don't like that.

p.s.
> And I don't think CDDA is about "realism" or "simulation" or other such bollocks,

It totally is, and always has been.

**Comment by Coolthulhu** [comment]
> Too many double negatives, do you mean zombie drops don't need nerfs? (they could definitely use some, they're everywhere, see below for fuller argument)

They are already nearly useless, except as a source of ingredients. Once in a blue moon you get a fitting drop of a hoodie or flame-resistant suit, but other than that there is pretty much no reason to treat zombie clothing as anything but not-yet-processed rags/leather/kevlar.

> That's a problem with crafting then, scavenging from houses and shops should (according to the general scenario) the dominant way to acquire equipment.

Houses don't have anywhere near enough actually useful gear because drops are padded with useless crap that takes more time to shift through than crafting own clothing. Zombie drops have one tiny advantage here: you can just use the 'V' listing to automatically shift (filter) through them and find the useful stuff, because you don't need to open every closet to find 30 panties, 10 different T-shirts, a silk top, a pair of regular pants (unfitting) and a mountain of other unprocessed ingredients per one single hoodie or jeans that you'd actually want to equip.

And then you STILL need that tailoring, because not well thought of mechanic of gear damage means supposedly tough clothing falls apart faster than skin underneath.

Meanwhile crafting avoids all that tedium by dumping it on the virtual character, which again is an unrealistic result of supposedly realistic mechanics clashing with the necessary limitations.
For many simulations, you need to go all the way before they start being realistic. Implementing sewing seems realistic because you can do it in real life, but ends up being less realistic because everyone becomes a tailor. Implementing real-life crafting times seems realistic, but then you'd end up with tailors having to go starving and dehydrated until you implement the ability to start now and finish later.
What matters is not the intent but the effect. If the effect isn't realistic, then the mechanic isn't either. And effect depends on all the metagame such as game balance and UI friendliness.

If we want scavenging fresh clothing to be more pronounced, it should be better than crafted or zombie stuff. And not "better" as in "this rare drop of superarmor can't be crafted", but "this hoodie has less fragile seams and is factory-crafted from good quality cloth rather than rags, +20% armor".

Filthy clothing should compete with crafted clothing. It doesn't - crafting takes less effort than washing. Most importantly less player effort.

**Comment by Night-Pryanik** [comment]
> It doesn't - crafting takes less effort than washing. 

It does - moreover, at lower levels washing surpasses crafting in terms of player effort. I see it like that: imagine player has zero to minimal tailoring skill and he has some good but filthy gear lying before him. What's better - wear filthy clothing right now (no effort at all) and get almost no penalties for immediate survival; or craft washboard, get some soap and find water (little effort, as wood, soap and water are almost ubiquitous) to wash that gear and remove all penalties; or spend time to find materials and spend time to train skills to be able to craft that gear (big effort)?

**Comment by Coolthulhu** [comment]
> It does - moreover, at lower levels washing surpasses crafting in terms of player effort.

I mean in the game, not real life.
In the game getting rags+needle+thread is far less of a hassle than getting washboard+soap+(enough) water.

**Comment by Night-Pryanik** [comment]
> In the game getting rags+needle+thread is far less of a hassle than getting washboard+soap+(enough) water.

I'd say they are roughly on par. 

- Rags = disassemble clothes or blanket or whatever.
Needle = have to find or craft (bone or wood variants).
Thread = disassemble rags or have to find.
- Washboard = need 2x4 and a tool with cutting quality.
Soap - have to find.
Water - have to find.

I'd say in both cases all items are very easy to find or craft. But filthy clothes have two advantages - first, medium to high level items are made not only from rags, so it may be harder to find required materials (leather, kevlar etc) and/or tools. Second - player can use them right now and right here, and wash them later, and the penalties (if any, according to gamestyle of given player) aren't so scary, while you have to - as I said earlier - spend time to find materials and spent time to train skills to be able to craft that gear.

**Comment by Coolthulhu** [comment]
> I'd say they are roughly on par.

Certainly not.
Rags are available in bulk in every house, needle is craftable at fabrication 1, thread is mass-produced (300 per) from strings on every house window.

Soap is semi-rare and a small city may well have no soap at all. There is no soap in wilderness at all.
Infinite water may be absent in cities, meaning you need funnels or lots of hauling.

It is so much less tedious, that I'd cut up fitting, undamaged filthy clothing just to craft its un-filthy versions before actually setting up washing.
The only exception is acid resistance items such as flame resistant suit, which is useful enough to actually consider washing (because acid res doesn't scale with damage to armor).

> But filthy clothes have two advantages - first, medium to high level items are made not only from rags, so it may be harder to find required materials (leather, kevlar etc) and/or tools.

You're very unlikely to find undamaged clothing that fits, meaning that you still need that crafting to repair everything into working shape. And even if you lucked out pretty hard and found a fitting, mostly-undamaged article of clothing, you'd need to repair it when it inevitably gets damaged.
And when you get those crafting skills, you can just make a comparable item, with less hassle than having to wash an existing one.

**Comment by keyspace** [comment]
> There is no soap in wilderness at all.

Preemptive: it can be said that "there are no zombies to drop filthy clothing either", but that is not true. With wandering on, some may stumble by.

**Comment by kevingranade** [comment]
>> Too many double negatives, do you mean zombie drops don't need nerfs? (they could definitely use some, they're everywhere, see below for fuller argument)

> They are already nearly useless, except as a source of ingredients. Once in a blue moon you get a fitting drop of a hoodie or flame-resistant suit, but other than that there is pretty much no reason to treat zombie clothing as anything but not-yet-processed rags/leather/kevlar.

In other words, it's a flavor change that also happens to nerf something that was already useless.  There's nothing wrong with that.
>> That's a problem with crafting then, scavenging from houses and shops should (according to the general scenario) the dominant way to acquire equipment.

> Houses don't have anywhere near enough actually useful gear because drops are padded with useless crap that takes more time to shift through than crafting own clothing. Zombie drops have one tiny advantage here: you can just use the 'V' listing to automatically shift (filter) through them and find the useful stuff, because you don't need to open every closet to find 30 panties, 10 different T-shirts, a silk top, a pair of regular pants (unfitting) and a mountain of other unprocessed ingredients per one single hoodie or jeans that you'd actually want to equip.

OK, how about an action-consuming action you can trigger* from V menu to 'search' a room and incorporate contents of containers?
If in-house drops are missing useful clothes that would reasonably be there, spawn lists can certainly be adjusted.
I'm far more interested in buffing home loot to resolve this issue, because both looting zombie drops and crafting clothing are utterly ludicrous ways to gear up.
* needs to be triggered manually so it doesn't rob moves from player without them intending it to.

> And then you STILL need that tailoring, because not well thought of mechanic of gear damage means supposedly tough clothing falls apart faster than skin underneath.

Or just loot more, there's an endless supply of new clothes.
> Meanwhile crafting avoids all that tedium by dumping it on the virtual character, which again is an unrealistic result of supposedly realistic mechanics clashing with the necessary limitations.
For many simulations, you need to go all the way before they start being realistic. Implementing sewing seems realistic because you can do it in real life, but ends up being less realistic because everyone becomes a tailor. Implementing real-life crafting times seems realistic, but then you'd end up with tailors having to go starving and dehydrated until you implement the ability to start now and finish later.

Those are all great things to have, crafting new clothes should honestly be a late-to-end game activity both from a time expenditure and skill point of view.  Instead of pushing forward to that though, you're defending the ridiculous status quo of the best options for early game gear being either zombie drops or tailoring.
Specifics aside, I think we agree that looting houses and shops need buffs to make them viable in the early game, and crafting clothing needs a niche further into the game.

> If we want scavenging fresh clothing to be more pronounced, it should be better than crafted or zombie stuff. And not "better" as in "this rare drop of superarmor can't be crafted", but "this hoodie has less fragile seams and is factory-crafted from good quality cloth rather than rags, +20% armor".

Sounds good, though that sounds like it might end up being complicated.  As I said above, we can also buff the availability of loot able clothing, and remove ui barriers to finding it.

> Filthy clothing should compete with crafted clothing. It doesn't - crafting takes less effort than washing. Most importantly less player effort.

I'm not convinced this is the case, filthy clothing is mostly trash, and the problem is crafting is orders of magnitude faster than it should be.

**Comment by Coolthulhu** [comment]
> OK, how about an action-consuming action you can trigger* from V menu to 'search' a room and incorporate contents of containers?

Reasonable workaround. Not perfect and not the proper "fix" (what we want here is essentially map memory+autosearch), but reasonable.

> If in-house drops are missing useful clothes that would reasonably be there, spawn lists can certainly be adjusted.

They would also need to be more likely to fit. One of the reasons not to ever scavenge for clothing is the 33% fit ratio.
But what the survivor is looking for is much different from "reasonable" clothing: pretty much all wool clothing is out, for example. No realistic reason, just stats.

> Or just loot more, there's an endless supply of new clothes.

Looting new clothing requires getting into not-yet-cleared areas, which requires - and consumes - good clothing. Stopping for a day and crafting good clothing is much more rewarding, safer and also less tedious (for the player).

>  Instead of pushing forward to that though, you're defending the ridiculous status quo of the best options for early game gear being either zombie drops or tailoring.

No, I'm attacking the idea of implementing actually anti-realistic mechanics, excusing them with realism, then deferring (forever) implementation of features necessary for said anti-realistic mechanics to make sense.
Status quo has the advantage: it isn't worse than half-implemented features that don't deliver "but would if".
As it stands, the filthy clothing would be detrimental. First make it good, then implement it. Not the other way.

Correct:
* Implement taking breaks from crafting without ruining the project
* Implement realistic crafting times
* Now that sources of clothing that aren't crafting are better than crafting, nerf those

Incorrect:
* Nerf weak options (zed clothing) to further limit and homogenize gameplay, reinforcing anti-realism
* Nerf crafting times causing uproar of "why do I need to cancel the whole crafting project to slap a zed kid?" from the players
* Finally, when everyone has lost hope for any sensible change, add taking breaks from crafting, which now kinda validates the changes before it

> Sounds good, though that sounds like it might end up being complicated. As I said above, we can also buff the availability of loot able clothing, and remove ui barriers to finding it.

Could be implemented simply as just another level of "negative item damage" over reinforced. That would be easy to implement, would not reward tailors more than they already are (because the "pristine" status would naturally go away as the item got damaged) and promote collecting good clothing.
It would be less work than buffing all those loot tables, which would require balancing good clothing with realistic clothing (most people don't stock up on hoodies).

**Comment by kevingranade** [comment]
> They would also need to be more likely to fit. One of the reasons not to ever scavenge for clothing is the 33% fit ratio.

Really? I thought fitting clothing was one of the least bad uses of tailoring.  What keeps that from working?

> Nerf weak options (zed clothing) to further limit and homogenize gameplay, reinforcing anti-realism

Do useful zombie drops almost never happen, or are they a critical source of clothing? You keep insisting that both are true, but that's not possible. 
> Nerf crafting times causing uproar of "why do I need to cancel the whole crafting project to slap a zed kid?" from the players
> Finally, when everyone has lost hope for any sensible change, add taking breaks from crafting, which now kinda validates the changes before it

Sure I'll buy that these two need to happen in the correct order, but that has nothing to do with making zombie drops actually useless.
> Sounds good, though that sounds like it might end up being complicated. As I said above, we can also buff the availability of loot able clothing, and remove ui barriers to finding it.

> Could be implemented simply as just another level of "negative item damage" over reinforced. That would be easy to implement, would not reward tailors more than they already are (because the "pristine" status would naturally go away as the item got damaged) and promote collecting good clothing.
It would be less work than buffing all those loot tables, which would require balancing good clothing with realistic clothing (most people don't stock up on hoodies).

They don't need to stock up, if you find one you're done for the time being.  Time to find the first instance of an item is critical, time to find additional instances is less so.
Also this isn't optional, no matter how much you nerf crafting and buff looted clothing, if the selection is bad we're back to tailoring being mandatory.

**Comment by keyspace** [comment]
Although current discussion ties into the PR, it has drifted way off. How about this be moved?

-----

> > They would also need to be more likely to fit. One of the reasons not to ever scavenge for clothing is the 33% fit ratio.

> Really? I thought fitting clothing was one of the least bad uses of tailoring. What keeps that from working?

It is indeed one of the best uses. However, [refitting is as hard as fixing a `max_madage()` item](https://github.com/CleverRaven/Cataclysm-DDA/blob/0b06b20edadb98f8d8c0299c0dc7f03aff67b95c/src/iuse_actor.cpp#L2461) (current code from PR #15424).

It is often the case that tailoring a fitting item is already possible, with high chance of success, whereas fitting an existing one still isn't.

Simple example: "pair of socks" is craftable at skill 0. However, refitting an unfit pair requires skilling-up to level 1 ("practice" action, trivial, a few hours and ~125 thread), at which point refitting becomes available, at 2% success / 2.4% failure chance.

TBH, it irks me that a character is always capable of tailoring clothing that fits. How about making that non-guaranteed?

-----

Preemptive reply to

> That will only make grinding even more necessary!

Grinding is possible due to current mechanic of experience gain: it is applied immediately. Only solution I've come up with is: _not_ applying it immediately. Say, drop it into a "buffer" instead, and limit stacking while doing so. Process the "buffer" as time goes by. "You think over your recent actions. Your survival skill has increased to 1!"

P.S. Found 3 relevant closed issues, will review and open a proposal issue if specifically this hasn't been discussed.

**Comment by Coolthulhu** [comment]
> Really? I thought fitting clothing was one of the least bad uses of tailoring. What keeps that from working?

Difficulty. I may have nerfed it a bit too much because it is surprisingly hard to strike the balance between "as hard as crafting" and "trivial to the point where refitting is just tedious keypresses".
Current refitting is as difficult as repairing very damaged clothing (uses same formula), which in turn is somewhat easier than crafting new clothing, except for very early levels (0-1). Generally faster than crafting new pieces (once again, except the trivial ones), but could be sped up a bit more.
Old refitting was trivial for all clothing at tailoring 2 or so and could be done in the middle of combat.

> Do useful zombie drops almost never happen, or are they a critical source of clothing? You keep insisting that both are true, but that's not possible.

Neither. Not sure where did you get the paradox.
Vast majority of zombie drops are useless, but some rare drops are useful enough to matter and make the zombie drops something more than trash that litters the screen.
Early backpacks, bags, acid-resistant clothing - things that aren't purely about protection from physical damage.

File: src/player.cpp [hunk 1]
```diff
@@ -4823,6 +4823,32 @@ dealt_damage_instance player::deal_damage(Creature* source, body_part bp, const
...
+    int sum_cover = 0, coverage = 0;
+    for( item &i : worn )
+    {
+        if( i.covers( bp ) )
+        {
+            coverage = i.get_coverage();
+            sum_cover += coverage;
+        }
+    }
+    const int infection_chance = dealt_dams.type_damage( DT_BASH ) + ( dealt_dams.type_damage( DT_CUT ) + dealt_dams.type_damage( DT_STAB ) ) * 4 * sum_cover / 100;
```
Comment (kevingranade) [inline src/player.cpp]: This scales based on how covered the body part is, regardless of whether the coverage comes from filty clothing or not.  It would make a lot more sense if it only summed up coverage from filthy clothing.  That would also eliminate the need to iterate over the worn clothing gain.
